### PR TITLE
Baseline rustdoc caching

### PR DESCRIPTION
**Pull request by mgr0dzicki**
I decided to place the code responsible for rustdoc caching in a separate file and write it in an object-oriented manner. I'm thinking also about extracting the logic of installing `cargo-semver-checks` in a similar way (in a separate PR). @obi1kenobi, what do you think about this approach?

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -85,6 +87,11 @@ async function installRustUp(): Promise<void> {
 }
 
 async function runCargoSemverChecks(cargo: rustCore.Cargo): Promise<void> {
+    // The default location of the target directory varies depending on whether
+    // the action is run inside a workspace or on a single crate. We therefore
+    // need to set the target directory explicitly.
+    process.env["CARGO_TARGET_DIR"] = "target";
```
Comment (obi1kenobi) [inline src/main.ts:93]: Did you test to make sure this does the right thing?

Internally, `cargo-semver-checks` uses [an explicit `--target-dir` option](https://github.com/obi1kenobi/cargo-semver-checks/blob/fe8ea52b8c461f433ff49f1c7ae47bc481689455/src/rustdoc_cmd.rs#L68-L79) when generating rustdoc, and I believe the explicit CLI option overrides the env var.

This would be great to have tests for, since it's something we could very easily accidentally break with serious consequences for our users.

File: src/rustdoc-cache.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import * as core from "@actions/core";
+import * as rustCore from "@actions-rs/core";
+
+export class RustdocCache {
+    private readonly cachePath;
+    private readonly cargo;
+    private __cacheKey = "";
+
+    constructor(cargo: rustCore.Cargo) {
+        this.cachePath = path.join("target", "semver-checks", "cache");
```
Comment (obi1kenobi) [inline src/rustdoc-cache.ts:16]: I wouldn't use a relative path here. Instead, I'd make the user supply a path in the constructor and then canonicalize it (make absolute + remove indirection) here before saving. Otherwise this is just a bug waiting to happen: caller constructs a cache, uses it in one place, then does a `cd` to another directory and 💥 bug.

File: src/rustdoc-cache.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                this.getCargoLocksHash(),
+                "semver-checks-rustdoc",
+            ].join("-");
+        }
+
+        return this.__cacheKey;
+    }
+
+    private getCargoLocksHash(): string {
+        const manifestPath = rustCore.input.getInput("manifest-path") || "./";
```
Comment (obi1kenobi) [inline src/rustdoc-cache.ts:56]: This is again a bug waiting to happen. "I'm running in the directory I'm supposed to check" is a reasonable assumption for the _action as a whole_ to make, but not for the _cache component_.

Please avoid the default and pass an explicit parameter in the constructor for this.

File: src/rustdoc-cache.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    private async getRustcVersion(): Promise<string> {
+        return await this.keyFromCommandCall("rustc", ["--version"]);
+    }
+
+    private async getCargoSemverChecksVersion(): Promise<string> {
+        return await this.keyFromCommandCall(this.cargo, ["semver-checks", "--version"]);
+    }
+
+    private async keyFromCommandCall(
+        command: string | rustCore.Cargo,
```
Comment (obi1kenobi) [inline src/rustdoc-cache.ts:72]: Let's not force this into being one function with two possible input types when it really wants to be two different functions. We even do an `if/else` based on the input type, so we aren't gaining much here -- we can get a better outcome by extracting the small snippet of common code into its own helper instead.

**Review by obi1kenobi — COMMENTED** [review]
Some minor suggestions for design and implementation improvements, so we avoid needing to debug things later...

File: src/main.ts [hunk 5]
```diff
@@ -129,7 +136,15 @@ async function run(): Promise<void> {
...
     await installCargoSemverChecks(cargo);
+
+    const cache = new RustdocCache(cargo);
+    const cacheFound = await cache.restore();
+
     await runCargoSemverChecks(cargo);
+
+    if (!cacheFound) {
+        await cache.save();
+    }
```
[the remark this answers is not in the document]
Comment (mgr0dzicki) [reply src/main.ts:147]: Well, maybe I didn't understand caching correctly. Is it possible that new crates will be added during running `cargo-semver-checks`? Since we hash all `Cargo.lock` files and use the hash as a part of the cache key, the cache won't be restored if any of them changes.

File: src/rustdoc-cache.ts (hunk 4 above)
Comment (mgr0dzicki) [reply src/rustdoc-cache.ts]: You're right, the code here wasn't good... I separated it into two functions, please have a look and say what do you think about it now.

File: src/main.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (obi1kenobi) [reply src/main.ts:147]: Hmm, are we guaranteed that newly added packages always cause a change in `Cargo.lock` even if they only use the same dependencies as other crates in the workspace? Will that continue to be the case forever, even if the `Cargo.lock` format changes? It seems to be on version 3 right now, so it could have a version 4 in the future.

I'd recommend looking into what `cache.save()` does if the cached data hasn't changed, and how long that takes. It's possible that the savings of not calling the function are minimal, and would be immediately wiped out (and then some) if we ever have to debug a problem related to this.

I trust you to look into the data and then use your best judgment on what to do here. You're the expert on this, having spent much more time looking into this than I have, so I trust you to make a good decision 👍

File: src/rustdoc-cache.ts [hunk 6]
```diff
@@ -61,32 +61,28 @@ export class RustdocCache {
...
+        const stdout = { s: "" };
+        await exec.exec("rustc", ["--version"], this.makeExecOptions(stdout));
+        return this.removeWhitespaces(stdout.s);
     }
 
     private async getCargoSemverChecksVersion(): Promise<string> {
-        return await this.keyFromCommandCall(this.cargo, ["semver-checks", "--version"]);
+        const stdout = { s: "" };
+        await this.cargo.call(["semver-checks", "--version"], this.makeExecOptions(stdout));
+        return this.removeWhitespaces(stdout.s);
```
Comment (obi1kenobi) [inline src/rustdoc-cache.ts:72]: Consider lifting functions like `makeExecOptions()` and `removeWhitespaces()` out of the class and just making them free functions in the module. They don't ever use `this`, and not having access to it makes them simpler to understand, less bug-prone in the future, and probably easier to optimize for the runtime as well.

**Review by obi1kenobi — COMMENTED** [review]
Definitely an improvement! Moving in the right direction 🚀

File: src/rustdoc-cache.ts (hunk 6 above)
Comment (mgr0dzicki) [reply src/rustdoc-cache.ts]: Good idea, changed!

File: src/rustdoc-cache.ts (hunk 2 above)
Comment (mgr0dzicki) [reply src/rustdoc-cache.ts]: Definitely a good idea! I changed it to be an argument of the constructor and used `path.resolve` to convert it to the absolute path. It should be much less prone to errors now...

File: src/rustdoc-cache.ts (hunk 3 above)
Comment (mgr0dzicki) [reply src/rustdoc-cache.ts]: I changed it similarly as above

File: src/main.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgr0dzicki) [reply src/main.ts:157]: `saveCache` provided by `@actions/cache` doesn't check whether the cached files changed, just uploads them. It's a very quick process, and even for large cache it would be a little fraction of the whole running time, so removing this `if` is not a problem. 

However, I still can't understand why we would need to change it, but this is probably my misunderstanding of how caching works in `cargo-semver-checks`. Could you explain it a little more? What do you mean by newly added packages? And can you give some concrete example of a situation when this approach might fail?

File: src/main.ts (hunk 1 above)
Comment (mgr0dzicki) [reply src/main.ts:94]: This is covered by `Test action inputs / Test input manifest-path against path containing space`: the `cargo-semver-checks` is run outside any workspace, with `--manifest-path="ref slice/Cargo.toml"`. Then, the default location would be `./ref slice/target/`, but it is being saved in `./target/`, as it should be. Otherwise, the test would fail, because `saveCache` throws an error if the supplied path does not exist.  

Do you think we can do any better? I'd be great to assert if the cache is being restored successfully (for example by looking for `baseline, cached` in the stdout), but as far as I know it is currently impossible to capture stdout of the whole action as a step in a workflow, so it would require changes in the action itself...

Comment (mgr0dzicki) [reply src/main.ts:94]: Oh, well, I just realized that relying on `saveCache` error is quite stupid at the moment, because it was called only once in the beginning as a result of the `if` statement we are discussing below... Removing the `if` will fix this issue, but maybe you have some better idea of how to test the caching?

Comment (obi1kenobi) [reply src/main.ts:94]: In the test workflow, could we just check after running the workflow that the target cache directory contains the files we expect and that they are not empty?

I'm less worried about the caching logic not uploading the files, and more worried about for example:
- a repeat of the bug where the baseline JSON is overwritten with the current JSON, where the cache directory ends up empty
- a `--workspace` run saving only the first / only the last package's baseline in the cache, and not the others

Also perhaps consider making the directory a bit more unique to our action. Our action shares the filesystem with everything else running in the same job, so it's possible that some other action also does something at `./target/` and may have an unfortunate interaction we may need to debug. Perhaps use `./semver-checks/target/` instead?

Zooming way out, here's why we're doing all this: we expect tons of people and projects to use this action, in all sorts of ways we can't predict ahead of time. There will be bugs and bad interactions with other tools, and we'll have to debug them — which is incredibly distracting, difficult, and time-consuming. Putting in additional work up front to systematically avoid 99.9% of those issues is necessary to ensure the success of the project: open-source projects die when they become too expensive to *maintain*.

File: src/main.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (obi1kenobi) [reply src/main.ts:157]: No problem, here's one scenario.

Say there's a workspace that contains multiple crates, and uses this action. Then, a new PR adds a new crate. This crate has no _new_ dependencies that weren't previously already used in the workspace — perhaps it has no dependencies _whatsoever_.

Does this modify the lockfile? In today's v3 lockfile format, yes: an entry for the new crate will be added to the lockfile.

_Must it_ modify the lockfile? No — no new information needed to be added, since this case could have been an "implicit default" case in the lockfile format: "if a crate doesn't have an entry in the lockfile, all its dependencies listed in `Cargo.toml` take the versions already specified in the lockfile." A hypothetical future lockfile format could add this behavior, since AFAIK the lockfile format doesn't ever promise "new packages added to the workspace will cause an edit to the lockfile." (There's also [Hyrum's Law](https://www.hyrumslaw.com/) to consider, and I'm sure the Rust team is aware of that law so it probably means this change won't actually be made...)

In that case, the hash and cache key would still match and will correctly _partially_ populate the cache — but the action won't upload the cached baseline for the _newly added crate in the PR_. Only when a new cargo-semver-checks version or rustc version is released will this situation get resolved, since that will reset the cache key.

This scenario isn't particularly likely (mostly because of the Rust team's competence and Hyrum's Law) so maybe we don't have to worry about it. But it's important to realize it's a possibility, seriously consider, and then make an explicit decision on ("worry about it / don't worry about it") rather than never thinking about it and making an unintentional random choice.

File: src/main.ts [hunk 7]
```diff
@@ -130,14 +131,23 @@ async function installCargoSemverChecks(cargo: rustCore.Cargo): Promise<void> {
     }
 }
 
+function getManifestDir(): string {
+    const manifestPath = rustCore.input.getInput("manifest-path") || "./";
+    return path.extname(manifestPath) ? path.dirname(manifestPath) : manifestPath;
+}
```
Comment (obi1kenobi) [inline src/main.ts:137]: nitpick: I would inline this function into `run()` because due to a subtlety, it actually isn't safe to call at any point outside of `run()`.

`run()` is the entry point of the action. It _knows_ what path `"./"` points to. This isn't true at any other point in the program.

I would also immediately canonicalize the path here. If the `"./"` default is used, we want to turn it into an absolute path ASAP or else we have the same problem as above.

File: src/main.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (obi1kenobi) [reply src/main.ts:157]: To be clear: it's _your call_ whether we leave the `if` or take it out. I was trying to help flesh out the pros and cons so you can have accurate input data to use when making the decision.

[the remark this answers is not in the document]
Comment (mgr0dzicki) [reply src/main.ts:157]: Thank you for such a detailed response! I was pretty sure that if a workspace contains multiple crates, a `Cargo.lock` is created separately for each of them, which is obviously wrong! Therefore I couldn't understand how the hash might not change if one new file is being hashed... Now I see that the difference is much more subtle, so it's better to save the cache again in any case.

However, there's one thing that is still bothering me. We're checking semver for library crates, isn't it very rare for them to contain `Cargo.lock` file?

File: src/main.ts (hunk 7 above)
Comment (mgr0dzicki) [reply src/main.ts:137]: Good idea, thank you! 

When I inline this function into `run()`, does it still make sense to canonicalize the path here? It's being resolved in `RustdocCache` constructor.

File: src/main.ts (hunk 1 above)
Comment (mgr0dzicki) [reply src/main.ts:94]: You're definitely right, `target` is quite a generic name, I'll change it to `semver-checks/target`.

I can easily prepare the test checking if the `target` directory exists and contains non-empty cache file (`registry-ref_slice-1_2_1.json` for the test repo). Is this sufficient? Or should I try to check for its validity in some way?

File: src/main.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (obi1kenobi) [reply src/main.ts:157]: > `saveCache` provided by `@actions/cache` doesn't check whether the cached files changed, just uploads them.

I'm not sure if this is true: based on the [code here](https://github.com/actions/cache/blob/main/src/saveImpl.ts#L40-L49) it seems like the action will recompute the cache key and will not upload if the key didn't change.

That means we have to include the cached rustdoc JSON files in the cache key.

> We're checking semver for library crates, isn't it very rare for them to contain `Cargo.lock` file?

Good point. Two things we should strongly consider in this case:
- Add a `prefix-key` argument like in [the Rust caching action](https://github.com/Swatinem/rust-cache#example-usage) to allow users to manually flush the cache if needed.
- Ensure that if the `Cargo.lock` file didn't exist _before_ the action, then it isn't taken into account when computing the cache key _after_ the action. This is because `Cargo.lock` will be created as part of most `cargo` operations in a workspace that doesn't contain it. We don't want those newly-generated `Cargo.lock` files going into the cache-saving cache key calculation because then we're guaranteeing that we never get any cache hits.

File: src/main.ts (hunk 7 above)
Comment (obi1kenobi) [reply src/main.ts:137]: Yes, I'd canonicalize everything immediately at the top of `run()`. That's the only place that has a consistent idea of where the action is running. Everything else should be path-independent since it can't guarantee what working directory it's running in.

File: src/main.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgr0dzicki) [reply src/main.ts:157]: > > `saveCache` provided by `@actions/cache` doesn't check whether the cached files changed, just uploads them.
> 
> I'm not sure if this is true: based on the [code here](https://github.com/actions/cache/blob/main/src/saveImpl.ts#L40-L49) it seems like the action will recompute the cache key and will not upload if the key didn't change.
> That means we have to include the cached rustdoc JSON files in the cache key.

This is the code of the GitHub `cache` action, which might be used in workflows, but not from the TypeScript code. Please notice that it calls `saveCache` function from the `@actions/cache` module (https://github.com/actions/cache/blob/main/src/saveImpl.ts#L59-L64) and this is the same function I'm using in `RustdocCache`. The code of `saveCache` can be found here https://github.com/actions/toolkit/blob/main/packages/cache/src/cache.ts#L165, and I'm pretty convinced it doesn't compare the key with the previous one, just uploads the files. Moreover, we cannot include the hash of rustdoc files in the key, because then we won't know the key at the beginning of the action and won't be able to download the cache... But we can just remove the `if` leaving bare `await cache.save()` and know the cache will change when it should. 

However I think we can do a bit better then that: download the cache, hash the cache directory, then run `cargo-semver-checks` and finally hash the cache directory again. Then we can do `cache.save()` only if these two hashes differ. I think it won't make much difference for small crates, but for these with really big rustdoc hashing the cache twice might be several seconds faster than uploading it unnecessarily. What do you think about it?

> > We're checking semver for library crates, isn't it very rare for them to contain `Cargo.lock` file?
> 
> Good point. Two things we should strongly consider in this case:
> 
>     * Add a `prefix-key` argument like in [the Rust caching action](https://github.com/Swatinem/rust-cache#example-usage) to allow users to manually flush the cache if needed.

Such an input already exists! I've called it `cache-key` and it defaults to `${{ github.job }}`. Do you think it should be described (or named) differently?

>     * Ensure that if the `Cargo.lock` file didn't exist _before_ the action, then it isn't taken into account when computing the cache key _after_ the action. This is because `Cargo.lock` will be created as part of most `cargo` operations in a workspace that doesn't contain it. We don't want those newly-generated `Cargo.lock` files going into the cache-saving cache key calculation because then we're guaranteeing that we never get any cache hits.

This isn't a problem, because the key is computed only once, before running `cargo-semver-checks`.

[the remark this answers is not in the document]
Comment (obi1kenobi) [reply src/main.ts:157]: Good catch, I missed the distinction between the actions implementation and the function called here.

> However I think we can do a bit better then that: download the cache, hash the cache directory, then run cargo-semver-checks and finally hash the cache directory again. Then we can do cache.save() only if these two hashes differ.

I think this is a great idea. Rustdoc JSON files can be hundreds of MB in size, and if we can avoid doing expensive compression and network operations with a quick hash check, we should.

Is it easy enough to implement now, or should we open an issue and tackle it in a follow-up PR?

> Such an input already exists! I've called it `cache-key` and it defaults to `${{ github.job }}`.

While I agree that `cache-key` is _capable_ of accomplishing the same thing, I don't think it's equivalent. Its stated intended role is to control which jobs share a cache, not do cache-busting. When used for cache-busting, let's keep in mind that most users won't be nearly as familiar with GitHub Actions as you and I have become, and are likely to make a mistake. For example, instead of changing `cache-key` from `${{ github.job }}` to `${{ github.job }}-1`, they might just change it to `1` and then be frustrated at the fact the cache is now shared across jobs.

In other words, we are not ensuring our users [fall into the pit of success](https://blog.codinghorror.com/falling-into-the-pit-of-success/) (highly recommended short read!)

This is why the [Rust caching action](https://github.com/Swatinem/rust-cache#example-usage) has so many different cache key settings: instead of saying "here's one cache key setting that technically can do everything," it says "here are many settings, each of which does exactly one thing and you can't accidentally misuse it to do anything else.

We should consider adopting a similar design and perhaps even the same naming terminology, with one key for cache-busting, another key for sharing caches across jobs, etc.

**Review by obi1kenobi — APPROVED** [review]
🚀
